### ISSUE TITLE
SSPI authentication is not working in wiltondb

### ISSUE THREAD

**Issue by Aravind2601** [wiltondb/wiltondb#37]
Hi @staticlibs,

we are trying to enable windows auth in pgadmin using SSPI auth but are getting error.
we have updated the below details in the pg_hba.conf and pg_ident.conf

pg_hba.conf:
![image](https://github.com/user-attachments/assets/2b147a07-e30b-4c4d-8e5d-caf3892767a0)
pg_ident.conf:
![image](https://github.com/user-attachments/assets/5cafc941-caeb-4d45-b666-c7c235e65f73)
Below is the error while we trying to log in:
![image](https://github.com/user-attachments/assets/becf03c4-ae95-415d-899a-62c1446bceb3)

Could you please help us if you aware any details on this?

**Comment by staticlibs** [comment]
@Aravind2601 , can you provide the details from the DB log about the SSPI failure - OS user, DB user and the map name should be listed there.

In your `pg_hba.conf` you need to specify the name of the map to use like this:

```
host    all             all          0.0.0.0/0               sspi map=sspi
```

where map name is taken from `pg_ident.conf`.

Also, in PgAdmin you need to enable Kerberos switch and also explicitly specify the name of the DB user to be selected after SSPI login is successful. The same user name must be used in `pg_ident.conf` something like this:

```
# MAPNAME       SYSTEM-USERNAME         PG-USERNAME
my_sspi_map alexk@DESKTOP-JS80C9O wilton
```

Also note, that TDS connections cannot be done on `pg_hba.conf` endpoints that are configured for `sspi` auth.

**Comment by Aravind2601** [comment]
@staticlibs ,
please find the below log and this after enable kerbos auth for server.
![image](https://github.com/user-attachments/assets/2702e755-74b1-44a5-9eee-63887aeb1b91)
Updated pg_hba.conf:
![image](https://github.com/user-attachments/assets/228ce1ec-78e1-4791-9e9a-a9064e6fb914)
Updated pg_ident_conf:
![image](https://github.com/user-attachments/assets/1e7363f0-f6b1-4b0d-a54c-01c6ffc87ee7)

**Comment by staticlibs** [comment]
@Aravind2601 , please use the OS name from error log in `pg_ident.conf`, I suppose it must be in Kerberos `user@DOMAIN` format instead of `domain\user` format.

**Comment by Aravind2601** [comment]
@staticlibs ,
I tried above one still getting some error .
pg_hba.conf:
![image](https://github.com/user-attachments/assets/ba73b451-09b8-4004-9d61-179077839c84)
pg_ident.conf:
![image](https://github.com/user-attachments/assets/47af9db4-afe1-4e8a-95e5-766c5a6ac434)
Log:
![image](https://github.com/user-attachments/assets/5550847e-d082-40ba-b8b6-45a5d538072b)

**Comment by staticlibs** [comment]
@Aravind2601 , on your screenshots the map name is different in pg_hba.conf and pg_ident.conf.

**Comment by Aravind2601** [comment]
@staticlibs Sry, actually I have updated base on your above response but now it's working fine after update Alex. 
![image](https://github.com/user-attachments/assets/8b87a1a6-b902-45ed-932f-6f668396fe23)
Thanks for the update .
One more question Alex, After enabling this SSPI auth, pgadmin is logging in with SSPI auth but we also need to use same windows auth for transfer, config and backup tool also . That you guys are update from your end?

**Comment by staticlibs** [comment]
@Aravind2601 , `wdb_transfer` works over TDS and already supports SSPI. `wdb_config` and `wdb_backup` do not support SSPI and it only makes sense to add it there if the same SSPI setup is used for both TDS and Postgres connections (as described in [this comment](https://github.com/wiltondb/wiltondb/issues/34#issuecomment-2286486330)) that is currently not possible.

**Comment by psrinivasa2** [comment]
Hi Alex @staticlibs , 
Our plan is to disable default Wilton user (may be by generating random unknown uuid password) , We will have only OS user which we will make owner for Wilton database. In this context, is it possible to connect wdb_backup or wdb_config tool? If not what would be the best solution for our scenario?

**Comment by psrinivasa2** [comment]
I could see a comment on this in below reference, we will check on it 
 
 https://github.com/wiltondb/wiltondb/issues/34#issuecomment-2286486330

**Comment by staticlibs** [comment]
@psrinivasa2 , I think the password user + pgpass file for Postgres connections will be the easiest to setup/maintain. OS user (created with "from windows") cannot be used for this though - such users are not allowed to perform password logins. Some ordinary user needs to be created for password logins (with pgpass).

**Comment by staticlibs** [comment]
@psrinivasa2 , about the SSPI support in wdb_backup/wdb_config - it is possible to implement, but DB setup in this case (user maps) is quite complex and these user maps must also be implemented for TDS - this is much bigger task than just adding pgpass support to wdb_backup/wdb_config.

**Comment by psrinivasa2** [comment]
Hi Alex @staticlibs , 
The main reason for us to create OS user is to avoid mentioning the password in files such as connection string file by considering security risks around it. Won't it be a security issue to mention the password in pgpass file? 
If we want to create the password user again, we end up maintain two user( password user, OS user) for all our 5000 sites. I will need to discuss on this with my senior management team.

**Comment by staticlibs** [comment]
@psrinivasa2 , 

> Won't it be a security issue to mention the password in pgpass file?

Can you clarify the security requirements regarding DB passwords handling you have in your environment?

> If we want to create the password user again, we end up maintain two user( password user, OS user) for all our 5000 sites.

As I understand it, your application is being moved from MSSQL and DB access from application will be done over TDS (where SSPI works with minimal friction).

Postgres protocol access is really necessary in Babelfish only for administrative tasks (both `wdb_backup` and `wdb_config`). Depending on the deployment setup, do you actually need remote access to DB over Postgres port? Postgres with [pg_hba.conf](https://www.postgresql.org/docs/15/auth-pg-hba-conf.html) has many options for networking and auth setup. For example to only allow local connections for a particular user, do not require password from a user coming from a particular networking address etc. Besides that, because OS user is planned to be used for TDS anyway, it may be possible to keep Postgres login disabled by default and only enable it (from OS user account) for administrative tasks and disable after that.

Assuming SSPI will be found the most convenient approach in your environment, I wonder whether it is feasible to do all the `pg_hba.conf` and `pg_ident.conf` setup required for SSPI for Postgres protocol for the large number of deployments you mention? I personally found that such setup (changing multiple conf files in Windows environment) to be cumbersome (if user needs to change any of these files) and error prone to user mistakes. So the idea for SSPI support on TDS was to not touch conf files at all. If you think that such setup is appropriate in your env, than getting support for it (`pg_hba.conf` + `pg_ident.conf`) for TDS connections would be a natural thing to do. This can be implemented (perhaps in more simplified form), though the task is much bigger that just adding pgpass file support.

**Comment by psrinivasa2** [comment]
Thanks for more details Alex @staticlibs. I need to look in more detail about the points you conveyed above, I will refer on next week once I back from my holidays and come back on this.

**Comment by psrinivasa2** [comment]
Meaning while helpful for us to get clarity on below point with step by step guide 
 
 _it may be possible to keep Postgres login disabled by default and only enable it (from OS user account) for administrative tasks and disable after that_

**Comment by staticlibs** [comment]
> Meanwhile helpful for us to get clarity on below point with step by step guide
> 
> _it may be possible to keep Postgres login disabled by default and only enable it (from OS user account) for administrative tasks and disable after that_

I referred to the possibility to use the OS user to control the access for secondary password user. For example, to reset its password on the command from the application as a part of some administrative task. This idea is a manual handling (not a Postgres feature) and have a number of assumptions (OS user being a superuser etc) that may or may not be possible.

**Comment by psrinivasa2** [comment]
Do you mean something like below,
 
1) install the WiltonDB which creates default user Wilton and Wilton password. 
2) create the OS user and disable Wilton user by generating random unknown uuid password 
3) when there is a need to use wdb_backup or wdb_config file, reset the Wilton user password to known password and use the backup and config tool 
4) reset the Wilton user password to unknown uuid password

**Comment by staticlibs** [comment]
@psrinivasa2 , yes, something like this. As an additional option to the possible other solutions like local-only Postgres user without password (through `pg_hba.conf`).

**Comment by psrinivasa2** [comment]
Thanks Alex @staticlibs , we will give a try on local only user too
